src/Elastic.Apm/Report/Serialization/LabelsJsonConverter.cs

```diff
@@ -23,16 +23,26 @@ public override void WriteJson(JsonWriter writer, Dictionary<string, string> lab
 					.Replace('"', '_'));
 
 				if (keyValue.Value != null)
-					writer.WriteValue(keyValue.Value.Truncate());
+				{
+					switch (keyValue.Value.Value)
+					{
+						case string strValue:
+							writer.WriteValue(strValue.Truncate());
+							break;
+						default:
+							writer.WriteValue(keyValue.Value.Value);
+							break;
+					}
+				}
 				else
 					writer.WriteNull();
 			}
 			writer.WriteEndObject();
 		}
 
-		public override Dictionary<string, string> ReadJson(JsonReader reader, Type objectType, Dictionary<string, string> existingValue,
+		public override LabelsDictionary ReadJson(JsonReader reader, Type objectType, LabelsDictionary existingValue,
 			bool hasExistingValue, JsonSerializer serializer
 		)
-			=> serializer.Deserialize<Dictionary<string, string>>(reader);
+			=> serializer.Deserialize<LabelsDictionary>(reader);
```

Comment:
```suggestion
		{
			if (reader.TokenType == JsonToken.Null)
				return null;

			if (reader.TokenType != JsonToken.StartObject)
				throw new JsonException($"expected {JsonToken.StartObject} but received {reader.TokenType}");

			var labels = new LabelsDictionary();

			while (reader.Read())
			{
				if (reader.TokenType == JsonToken.EndObject)
					break;

				var property = (string)reader.Value;
				reader.Read();

				switch (reader.TokenType)
				{
					case JsonToken.Integer:
						labels.InnerDictionary.Add(property, (long)reader.Value);
						break;
					case JsonToken.Float:
						labels.InnerDictionary.Add(property, (double)reader.Value);
						break;
					case JsonToken.String:
						labels.Add(property, (string)reader.Value);
						break;
					case JsonToken.Boolean:
						labels.InnerDictionary.Add(property, (bool)reader.Value);
						break;
					default:
						throw new JsonException(
							$"Expected {JsonToken.Integer}, {JsonToken.Float}, {JsonToken.String} or {JsonToken.Boolean} "
							+ $"but received {reader.TokenType}");
				}
			}

			return labels;
		}
```

Reply:
One question that came to mind - do labels need to support dates? Understand that they will be serialized as a JSON string, but wondering if deserialization has to consider `JsonToken.Date`

Reply:
> do labels need to support dates? 

No, my understanding is that no other agent supports dates. It's only string, numbers, and bool. I guess technically we could support dates, but so far I followed other agents and those don't support it. Before we add it I'd definitely see if there is feedback and need for it at al.

>  wondering if deserialization has to consider `JsonToken.Date`

Since the de-serialization logic is not used in production and I feel date isn't needed to be supported I think it's fine to ignore it. As discussed over zoom, the reason I'm not very worried about the de-serialization logic is that the agent only uses the serialization logic to go from C# objects to JSON, but not the other way, so if de-serialization is important at al, then it's maybe in some of our internal tests scenarios (but currently that's not even a case based on what we discussed (Mock APM Server e.g. does not use this)).